src/Receiver.cc

```diff
@@ -90,83 +108,91 @@ Receiver::handleDataPacket(Driver::Packet* packet, IpAddress sourceIp)
 {
     Protocol::Packet::DataHeader* header =
         static_cast<Protocol::Packet::DataHeader*>(packet->payload);
-    uint16_t dataHeaderLength = sizeof(Protocol::Packet::DataHeader);
     Protocol::MessageId id = header->common.messageId;
 
+    bool needSchedule = false;
+    bool messageComplete;
     MessageBucket* bucket = messageBuckets.getBucket(id);
-    SpinLock::Lock lock_bucket(bucket->mutex);
-    Message* message = bucket->findMessage(id, lock_bucket);
-    if (message == nullptr) {
-        // New message
-        int messageLength = header->totalLength;
-        int numUnscheduledPackets = header->unscheduledIndexLimit;
-        {
-            SpinLock::Lock lock_allocator(messageAllocator.mutex);
+    Message* message;
+    {
+        // Scoped critical section guarded by MessageBucket::mutex; this ensures
```

Comment:
Athough it's unlikely, it's possible this message could be destroyed following this critical section meaning the rest of the method following the critical section would be accessing an invalid `message`.

Reply:
I see. So this is the real reason why the bucket mutex has to be held till the end of the method. It worries me that safe memory reclamation depends on both the bucket mutex and schedulerMutex (e.g., you don't acquire a bucket mutex in `trySendGrants()`). For now, I have added some more docs to make it clear.